Add fast path to extract_band and bandjoin for single uchar, up to ~60% faster for 1 and 3 channels

lovell opened the pull request:
Such images are common and using these two functions for single channel remove/add is also common.

Removing the inner `z` loop allows the compiler to auto-vectorise the outer `x` loop.

Here are some `callgrind` stats for extracting a single channel from a 3 channel 1.4 megapixel image:

Before:
```
28,466,438 (19.51%)  ../libvips/conversion/extract.c:vips_extract_band_buffer [/usr/local/lib/x86_64-linux-gnu/libvips.so.42.17.0]
```
After:
```
 8,568,560 ( 6.80%)  ../libvips/conversion/extract.c:vips_extract_band_buffer [/usr/local/lib/x86_64-linux-gnu/libvips.so.42.17.0]
```

...and some stats for joining a channel to the same 3 channel image:

Before:
```
71,131,060 (36.56%)  ../libvips/conversion/bandjoin.c:vips_bandjoin_buffer [/usr/local/lib/x86_64-linux-gnu/libvips.so.42.17.0]
```

After:
```
51,234,183 (29.32%)  ../libvips/conversion/bandjoin.c:vips_bandjoin_buffer [/usr/local/lib/x86_64-linux-gnu/libvips.so.42.17.0]
```

I haven't done it here, but it might be worth adding a fast path for 3 channel joining (`ips == 3`) as well as 1 channel, as RGB+A is a common scenario. Happy to add to this PR or as a follow-up.

There is already a fast path for `vips_bandjoin_const` and this approach keeps things consistent.

https://github.com/libvips/libvips/blob/07d3a53f0d29fa5e611ed372b879117bb53af3de/libvips/conversion/bandjoin.c#L328

jcupitt commented on libvips/conversion/bandjoin.c:
```diff
@@ -116,15 +116,27 @@ vips_bandjoin_buffer( VipsBandarySequence *seq,
-				q1[z] = p1[z];
+		if( ips == 1 ) {
+			for( x = 0; x < width; x++ ) {
+				q1[0] = p1[0];
```
You could swap this for `p1[x]` and get rid of the `p1 += 1` below, it should save a little more.

jcupitt commented on libvips/conversion/extract.c:
```diff
@@ -355,17 +355,27 @@ vips_extract_band_buffer( VipsBandarySequence *seq,
+	if( ops == 1 ) {
+		for( x = 0; x < width; x++ ) {
+			q[0] = p[0];
+			p += ips;
```
And here it could be `q[x]` and no `+= 1`.

jcupitt commented:
Nice! I added two possible tiny speedups (I've not actually timed them).

lovell commented:
I've added John's suggestions. They don't appear to offer any additional performance improvement so I guess the compiler had already realised the two variables were in step.

I also added the fast path for joining 3 channels, which makes the common RGB+A scenario much faster:

Before:
```
71,131,060 (36.56%)  ../libvips/conversion/bandjoin.c:vips_bandjoin_buffer [/usr/local/lib/x86_64-linux-gnu/libvips.so.42.17.0]
```
After:
```
22,813,791 (15.59%)  ../libvips/conversion/bandjoin.c:vips_bandjoin_buffer [/usr/local/lib/x86_64-linux-gnu/libvips.so.42.17.0]
```

kleisauke commented on libvips/conversion/bandjoin.c:
```diff
@@ -116,15 +116,38 @@ vips_bandjoin_buffer( VipsBandarySequence *seq,
+				q1 += ops;
+			}
+
+			q += 1;
```
```suggestion
			q += ips;
```
To make it consistent with the `else` statement, although I'm not sure if that would cause a different codegen.

kleisauke commented on libvips/conversion/bandjoin.c:
```diff
@@ -116,15 +116,38 @@ vips_bandjoin_buffer( VipsBandarySequence *seq,
+				q1[1] = p1[1];
+				q1[2] = p1[2];
+
+				p1 += 3;
```
```suggestion
				p1 += ips;
```
ditto.

kleisauke commented on libvips/conversion/bandjoin.c:
```diff
@@ -116,15 +116,38 @@ vips_bandjoin_buffer( VipsBandarySequence *seq,
+			}
 
-		q += ips;
+			q += 3;
```
```suggestion
			q += ips;
```
ditto.

kleisauke reviewed (approved):
Great! I left a few minor comments inline, but I'm not sure if that would affect performance.

kleisauke commented:
FWIW, I see no difference in assembly output when I apply my suggestions locally on both GCC 12.2.1 and Clang 15.0.7. Tested with `CFLAGS="-O3 -save-temps"`.

kleisauke commented:
Thank you! :+1:
